`FinalityView`, a `DbReader` wrapper to represent a finalized state

mzabaluev opened the pull request:
### Description

To provide the API view on the finalized blockchain state, add a wrapper that can be set with a `LedgerInfoWithSignatures` to designate and validate some ledger version as the "latest".

### Test Plan

Test behavior of the customized `DbReader` methods of the `FinalityView` wrapper before and after the finalized ledger info is set.

mzabaluev committed:
Remove calls to AptosDataBuilder::is_async

Some tests still had these.

mzabaluev committed:
aptos-db: remove reversion helper method

AptosDB::commit_reversion_ledger_info
does the same thing as commit_ledger_info,
just use that method instead.

mzabaluev committed:
storage-interface: add FinalityView

This wrapper is meant to represent the finalized state
for the API.

mzabaluev committed:
storage-interface: tests for FinalityView

mzabaluev committed:
storage-interface: blanket impl DbReader for Arc

0xmovses commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,118 @@
+
+impl<Db: DbReader> FinalityView<Db> {
+    /// Updates the information on the latest finalized block's ledger.
+    pub fn set_finalized_ledger_info(&self, ledger_info: LedgerInfoWithSignatures) -> Result<()> {
```
accept `blockheight` here instead of `ledger_info` then fetch the blockinfo from the DB reader and populate it inside this method.

0xmovses commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    fn get_latest_ledger_info_option(&self) -> Result<Option<LedgerInfoWithSignatures>> {
+        let fin_ledger_info = self.finalized_ledger_info.read().unwrap();
```
Return error here instead of `unwrap()`? I know this means the Rw would be poisned anyway.

0xmovses reviewed (commented):
Left some comments for myself to implement.

0xmovses committed:
update set ledger info method, update tests

0xmovses committed:
update tests update set method

mzabaluev commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -82,26 +100,86 @@ mod tests {
+
+        // Capture the block event once
+        let (_start_ver, end_ver, block_event) =
+            view.get_block_info_by_height(blockheight).unwrap();
```
I'd rather use a clone of the mock to get the expected values than go through the object under test, just to prevent any masked problems if we implement more view methods.

0xmovses replied:
Will update.

mzabaluev committed:
storage-interface: fixed up finality_view tests

No randomness allows us to verify the hash.

mzabaluev committed:
storage-interface: fix DbReader impl for Arc

get_read_delegatee needs to deref to get the vtable
for the pointee type.

mzabaluev committed:
storage-interface: rename setter for FinalityView

FinalityView::set_finalized_block_height better describes
the argument.

mzabaluev committed:
storage-interface: remove a dbg printout

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+        // Sanity checks: finalization should not be set on an empty database,
+        // the finality version should not exceed the latest committed.
+        match self.reader.get_latest_state_checkpoint_version()? {
+            None => return Err(AptosDbError::Other("no ledger states to finalize".into())),
```
So, we _are_ relying on checkpoints?

mzabaluev replied:
All blocks in _opt_ should be committed to storage by the opt executor, so it's the best sanity check I could come up with that does not require explicit sync-ups from the opt state.
In another change @0xmovses added the code above to fetch block info; I'll check if it already ensures that we get it from a committed block and this check is redundant.

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+            },
+        }
+        let ledger_info = LedgerInfo::new(block_info, HashValue::zero()); // we don't use consensus_data_hash
+        let aggregate_signature = aggregate_signature::AggregateSignature::empty(); // we don't use
```
Maybe provide an extended comment here.

l-monninger replied:
Explain why we don't use.

mzabaluev replied:
This is one reason why, with my aptos-core hat on, I wanted to pass `LedgerInfoWithSignatures` as input to this API and synthesize it at the maptos-fin-view level where we know we can set these fields with trivial values.

mzabaluev replied:
Thanks for picking on this. I think I have figure out a proper way to retrieve the ledger info; it's the latest one committed for the block's epoch.

mzabaluev replied:
The idea did not work out without modifying more of `DbReader`, so we'll have to make do with the current method.
I have added a comment describing the Movement-specific construction of `LedgerInfoWithSignatures`.

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+        Ok(version)
+    }
+
+    // TODO: override any other methods needed to maintain the illusion.
```
What are these methods ATM?

mzabaluev replied:
Anything in the API that does not use `get_latest_ledger_info` or construct a state view around the latest version, but takes a version as a parameter, would be able to get objects from non-finalized state versions. If this is an issue, we need to override the relevant methods and error out on versions that are newer than the finalized one.

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+    use crate::mock::MockDbReaderWriter;
+
+    #[test]
+    fn test_get_latest_ledger_info() {
```
Question, as a team do we prefer tests with `Result` or with `unwrap`. I like `Result` for two reasons:
1. Tests often become defacto documentation and people copy examples from them.
2. Verbosity.

mzabaluev committed:
Revert "fix: debugging transaction_by hash issue."

This reverts commit 7f3e0e5c93d2b208f421445e3fd7ecdad8d0a4dd.

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+
+/// A wrapper over [`DbReader`], representing ledger at the end of a block
+/// as the latest ledger and its version as the latest transaction version.
+pub struct FinalityView<Db> {
```
I feel we may want to call this `FinView` since we are referring specifically to the `fin` concept outlined here: https://github.com/movementlabsxyz/rfcs/tree/main/0018-m2-state-finality-modes

mzabaluev replied:
I was naming things with my aptos-core hat on, so I used more general terms in this code base.

l-monninger reviewed (changes requested):
Tests are passing. Comments are for awareness and aesthetic changes.

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+        // Set the finalized ledger info
+        view.set_finalized_block_height(blockheight).unwrap();
+
+        // Capture the block event once
```
Do we know how the `BlockEvent` is used by the API precisely--if at all?

mzabaluev replied:
There is [a comment](https://github.com/movementlabsxyz/aptos-core/blob/ea91067b81f9673547417bff9c70d5a2fe1b0e7b/storage/aptosdb/src/db/include/aptosdb_reader.rs#L663) stating the developers would rather return `BlockInfo` from this method, but we make do with information from this structure.

l-monninger commented on storage/storage-interface/src/finality_view.rs:
```diff
@@ -0,0 +1,159 @@
+        let version = view.get_latest_state_checkpoint_version().unwrap();
+        assert_eq!(version, None);
+        view.set_finalized_block_height(1).unwrap();
+        let version = view.get_latest_state_checkpoint_version().unwrap();
```
I may be a bit behind the ball here on our latest means of handling things.  But...

Is there a reason we shouldn't have a test here that actually checks whether the state made available corresponds to what we would expect for a given ledger version/block height? 

I suppose since it seems we are still always working off of checkpoints this is effectively checked by the original `aptos-core` tests.  But, it would be good to perhaps link to where those are.

If we change this in the future, should we include checks at this level that would make sure this fin view actually marries up with the proper `StateView`?

mzabaluev replied:
I'll add a test specifically for a state view created around `FinalityView` to check that the states are only available to the finality version.

mzabaluev replied:
The relevant view is the one obtained by the `Context::latest_state_view*` methods.
I wrote a stub of a test for it, but proper testing would need more support from the mock.

mzabaluev replied:
The test in https://github.com/movementlabsxyz/movement/pull/91/commits/3541365d18aa36102464b7d1dd8079870f9f2b54 tests that it works on a real populated database.

mzabaluev committed:
storage-interface: finality_view tests ret Result

Use ? for terser, idiomatic code that can be cribbed for production.

mzabaluev committed:
storage-interface: latest state vew test

Draft a test for FinalityView, checking that the latest_state_checkpoint_view
extension method gets us the view into finalized state.
StateView support in the mock is needed to complete the test.

mzabaluev committed:
storage-interface: clean up and comment

In FinalityView::set_finalized_block_height, remove the outdated
sanity check, add comments explaining why we can use the empty hash
and no signatures specifically for Movement.

mzabaluev committed:
storage: improve FinalityView set method

For set_finalized_block_height, retrieve ledger info from the
corresponding epoch or use the latest.

mzabaluev committed:
storage: more FinalityView testing

Exercise the DbStateView instance returned from
latest_state_checkpoint_view. No testing yet if the state
matches expectations for the finalized version.

mzabaluev committed:
Revert "storage: improve FinalityView set method"

This reverts commit 1731cce36e3810d4034240deeefffd3745f8dedc.

mzabaluev committed:
api: expose AccountsApi::get_account_inner

Access to account state is convenient to test with.
